Add a proper `File` class

flo12356 opened the pull request:
### Description

Add class `File` as part of the wrapper for littlefs.

Fixes parts of #273
Fixes parts of #186
Fixes parts of #209

flo12356 committed:
Add file class

flo12356 committed:
Restructure file class

PatrickKa committed:
Refactor LfsWrapper a bit

flo12356 committed:
WIP try to fix errors in lfsWrapper.test.cpp

PatrickKa committed:
Add NewOutcome.test.cpp

PatrickKa committed:
Format Outcome.hpp and print error messages before aborting

PatrickKa committed:
Format file system if mounting does not work

PatrickKa committed:
Add -g to debug flags in CMake presets

PatrickKa committed:
Change debug optimization level from -Og to -O0

PatrickKa committed:
Turn lfs wrapper test into normal executable

The TEST_CASE()s and other magic from Catch2 made it harder to debug.

PatrickKa committed:
Fix move construction and assignment of File

This requires addinge the members path_ and openFlags_.

PatrickKa committed:
Add File.isOpen_

flo12356 committed:
Move template functions to ipp

flo12356 committed:
Add LfsWrapper Unit tests

flo12356 committed:
Add flag check befor read/write

PatrickKa committed:
Use statically allocated file buffer

The normal lfs_file_open() uses malloc() to allocate the file buffer.
Since we do not want to use the heap, we use lfs_file_opencfg() and
provide a statically allocated buffer instead.

PatrickKa committed:
Disable warning on missing field initializers

When using designated initializes to initialize a struct, it is
perfectly valid not to specify all fields. The unspecified fields are
initialized with default member initializers where provided, and empty
list-initialization otherwise.

PatrickKa committed:
Add fs::ErrorCode::unsupportedOperation

PatrickKa committed:
Remove [[nodiscard]] from function definitions

PatrickKa committed:
Use almost always auto in LfsWrapper.test.cpp

PatrickKa committed:
Make flags in LfsWrapper unsigned

Flag varibales are one of the few cases where unsigned types should be
used because they do not represent a real quantity/number but just a
bunch of states/bits.

PatrickKa committed:
Rename storage to memory

flo12356 committed:
WIP: Add move function

flo12356 committed:
WIP: Replace move with swap

PatrickKa committed:
Revert "WIP: Replace move with swap"

This reverts commit 1d0bd2f8dd40c48ab1a5b2d1b6f76eed6b954481.

In this special case the Swap() function is an abomination, and the
previous version with the Move() function is better.

PatrickKa committed:
Refactor `fs::File`

- Rename `Move()` to `MoveConstructFrom()`
- Make the class const-correct
- Remove `std::cout` and `iostream`
- Make cosmetic changes

flo12356 committed:
Fix rebase error

flo12356 committed:
Add internal read write functions for LfsWrapper

PatrickKa committed:
Fix target_link_libraries of FileSystem

PatrickKa committed:
Remove LfsWrapperExecutable and NewOutcome tests

They were used to test stuff during development but are now obsolete.

PatrickKa committed:
Fix target_link_libraries of LfsWrapper test

PatrickKa committed:
Fix includes of LfsWrapper test

PatrickKa commented on Sts1CobcSw/FileSystem/LfsWrapper.ipp:
```diff
@@ -0,0 +1,22 @@
+
+#include <Sts1CobcSw/FileSystem/LfsWrapper.hpp>
+
+#include "Sts1CobcSw/FileSystem/ErrorsAndResult.hpp"
```
Only the corresponding .hpp file may be included in an .ipp file, so this line must be removed.

flo12356 replied:
Fixed in commit https://github.com/SpaceTeam/STS1_COBC_SW/pull/287/commits/e6efcca7d93be7354637641ebe498b194f0e4f44

PatrickKa commented on Sts1CobcSw/Outcome/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-target_link_libraries(Sts1CobcSw_Outcome INTERFACE rodos::rodos)
+# target_link_libraries(Sts1CobcSw_Outcome INTERFACE rodos::rodos)
```
I guess this is related to the whole `USE_NO_RODOS` issue. We'll probably need two `Outcome` targets: one with and one without Rodos. The unit tests can then link to the one without Rodos while the rest uses the "normal" version.

flo12356 replied:
Readded in https://github.com/SpaceTeam/STS1_COBC_SW/pull/287/commits/b395f28ac8899dddc896183bcb19de1f1184792b

PatrickKa reviewed (changes requested):
- `add_compile_definitions(USE_NO_RODOS)` in the top-level CML file seems wrong
- `Read/Write()` should be like `fram::ReadFrom/WriteTo()`, i.e. they should take byte spans:

  ~~~c++
  template<std::size_t extent>
  [[nodiscard]] auto Read(std::span<Byte, extent> data) const -> Result<int>;
  template<std::size_t extent>
  [[nodiscard]] auto Write(std::span<const Byte, extent> data) -> Result<int>;
  ~~~

- Rename `Read/WriteInternal()` to just `Read/Write()`. The private functions have different parameters, so there is no reason to give them different names for disambiguation (function overloading FTW).
- Use the same order for the public and private `Read/Write()` functions

flo12356 committed:
Fix includes for LfsWrapper.ipp

flo12356 committed:
Refactor LfsWrapper read/write function

flo12356 committed:
Fix USE_NO_RODOS compile define for unit tests

flo12356 committed:
Move USE_NO_RODOS define

flo12356 committed:
Add lockfile for LfsWrapper

flo12356 committed:
Add unit test for LfsWrapper lockfile

PatrickKa committed:
Reduce max. path length to 30 characters

The longest path should be "/programs/65536.zip.lock" which is only 25
characters long, so we can save quite a bit of memory when we reduce the
max path length from the default 256 to just 30.

PatrickKa committed:
Refactor `LfsWrapper.[ch]pp`

- Always pass paths as `Path` aka `etl::string` and use `c_str()` to
  ensure that the littlefs functions get null-terminated strings
- Move member variables which are only used in `CreateLockFile()` to
  that function

PatrickKa committed:
Get rid of `USE_NO_RODOS` and rework `LfsWrapper.test.cpp`

Eventually, the `File` class will need semaphores and therefore require
the main function from Rodos. I converted it now, because that allows me
to get rid of this `USE_NO_RODOS` nonsense in `Outcome.hpp` too.

PatrickKa committed:
Link correct libraries for `Sts1CobcSw_FramSections`

PatrickKa reviewed (changes requested):
- Make `CreateLockFile()` and `Close()` `const`
- In `Open()` when calling `CreateLockFile()` either use `OUTCOME_TRY(file.CreateLockFile());` or create (and return) a new, specific error code for this, like `fileLocked`
- Instead of removing the lock file in `MoveConstructFrom()` by calling `Close()` and creating it again later, we should add a private `CloseAndKeepLockFile()` function that we call instead. `Close()` should then also call this function and additionally remove the lock file.

flo12356 committed:
Add closeAndKeepLockFile for LfsWrapper

flo12356 committed:
Fix const correctness in LfsWrapper

PatrickKa committed:
Use `File::CloseAndKeepLockFile()` in `File::Close()`

PatrickKa committed:
Add `ErrorCode::fileLocked`
